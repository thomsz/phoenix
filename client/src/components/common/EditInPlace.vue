<template>
  <div class="edit-in-place">
    <form
      v-if="edit"
      :name="`edit_in_place__${name}`"
      @submit.prevent="submit"
    >
      <textarea
        v-if="isTextarea"
        class="w-full p-2 bg-transparent"
        v-model="innerValue"
        :name="name"
        :rows="textareaRows"
        @focus="lockEdit"
        @focusout="submit"
        @mouseleave="toggleUnlockedEdit"
      >
      </textarea>
      <input
        v-else
        class="text-slate-400 w-full h-full px-2 bg-transparent"
        v-model="innerValue"
        :name="name"
        :type="type"
        @focus="lockEdit"
        @focusout="submit"
        @mouseleave="toggleUnlockedEdit"
      />
    </form>
    <div
      v-else
      class="p-2 text-slate-500"
      data-testid="interactive-element"
      :class="{ 'whitespace-pre-wrap': isTextarea }"
      @click="toggleEdit"
      @mouseover="toggleEdit"
    >
      <slot>
        {{ modelValue || '-' }}
      </slot>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'EditInPlace',

  props: {
    name: {
      type: String,
      required: true
    },

    type: {
      type: String,
      default: 'text'
    },

    modelValue: {
      type: String,
      required: true
    }
  },

  data () {
    return {
      edit: false,
      editLocked: false,
      innerValue: this.modelValue
    }
  },

  computed: {
    isTextarea (): boolean {
      return this.type === 'textarea'
    },

    textareaRows (): number {
      const value = this.innerValue
      const newLines = (value.match(/\n/g) || []).length
      const ROW_OFFSET = 1
      return newLines + ROW_OFFSET
    }
  },

  methods: {
    submit (): void {
      this.toggleEdit()

      if (this.innerValue === this.modelValue) return
      this.$emit('update:modelValue', this.innerValue)
    },

    lockEdit (): void {
      this.editLocked = true
    },
    
    unlockEdit (): void {
      this.editLocked = false
    },

    toggleEdit (): void {
      if (this.edit) {
        this.unlockEdit()
      }

      this.edit = !this.edit
    },

    toggleUnlockedEdit (): void {
      if (this.editLocked) return
      this.toggleEdit()
    }
  }
})
</script>

<style scoped>
.edit-in-place,
.edit-in-place > * {
  @apply w-full h-full flex items-center;
}
</style>
