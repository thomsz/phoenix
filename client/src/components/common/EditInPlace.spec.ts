import { faker } from '@faker-js/faker'
import { shallowMount } from '@vue/test-utils'

import EditInPlace from '@/components/common/EditInPlace.vue'

describe('EditInPlace', () => {
  const mountingOptions = {
    props: {
      name: faker.random.word(),
      modelValue: faker.random.words()
    }
  }

  ;['click', 'mouseover'].forEach(event => {
    it(`should toggle edit: ${event}`, async () => {
      const wrapper = shallowMount(EditInPlace, mountingOptions)
      const interactiveElementWrapper = wrapper.find('[data-testid="interactive-element"]')
      expect(wrapper.find('form').exists()).toBe(false)
      expect(interactiveElementWrapper.exists()).toBe(true)

      await interactiveElementWrapper.trigger(event)
      expect(wrapper.find('form').exists()).toBe(true)
      expect(wrapper.find('[data-testid="interactive-element"]').exists()).toBe(false)

      await wrapper.find('form').trigger('submit')
      expect(wrapper.find('form').exists()).toBe(false)
      expect(interactiveElementWrapper.exists()).toBe(true)
    })
  })

  ;[
    ['text', 'input'],
    ['textarea', 'textarea']
  ].forEach(([type, element]) => {
    it(`should render form element: ${type}`, async () => {
      const wrapper = shallowMount(EditInPlace, {
        ...mountingOptions,
        props: {
          ...mountingOptions.props,
          type,
          name: 'fake_name',
          modelValue: 'Fake value'
        }
      })

      expect(wrapper.html()).toMatchSnapshot()

      await wrapper.setData({ edit: true })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it(`should exit edit mode: ${type}`, async () => {
      const wrapper = shallowMount(EditInPlace, {
        ...mountingOptions,
        props: {
          ...mountingOptions.props,
          type
        },
        data: () => ({
          edit: true
        })
      })

      await wrapper.find(element).trigger('mouseleave')
      expect(wrapper.find('form').exists()).toBe(false)
      expect(wrapper.find('[data-testid="interactive-element"]').exists()).toBe(true)
    })

    it(`should not exit edit mode when field is focused: ${type}`, async () => {
      const wrapper = shallowMount(EditInPlace, {
        ...mountingOptions,
        props: {
          ...mountingOptions.props,
          type
        },
        data: () => ({
          edit: true
        })
      })

      const elementWrapper = wrapper.find(element)
      await elementWrapper.trigger('focus')
      await elementWrapper.trigger('mouseleave')
      expect(wrapper.find('form').exists()).toBe(true)
      expect(wrapper.find('[data-testid="interactive-element"]').exists()).toBe(false)
    })

    ;[
      ['changed value', faker.random.words(), true],
      ['unchanged value', mountingOptions.props.modelValue, false]
    ].forEach(([descriptor, submittedValue, shouldUpdateModelValue]) => {
      it(`should submit on blur: ${type} - ${descriptor}`, async () => {
        const wrapper = shallowMount(EditInPlace, {
          ...mountingOptions,
          props: {
            ...mountingOptions.props,
            type
          },
          data: () => ({
            edit: true
          })
        })

        const elementWrapper = wrapper.find(element)
        elementWrapper.setValue(submittedValue)
        await elementWrapper.trigger('focusout')

        expect(wrapper.find('form').exists()).toBe(false)
        expect(wrapper.find('[data-testid="interactive-element"]').exists()).toBe(true)
        expect(wrapper.emitted('update:modelValue')?.[0][0]).toBe(shouldUpdateModelValue ? submittedValue : undefined)
      })
    })
  })

  Array.from({ length: 3 }, (_, index) => {
    const ROW_OFFSET = 1
    return index + ROW_OFFSET
  }).forEach((rowCount) => {
    it(`should set number of textarea value rows: ${rowCount}`, () => {
      const wrapper = shallowMount(EditInPlace, {
        ...mountingOptions,
        props: {
          ...mountingOptions.props,
          type: 'textarea',
          modelValue: faker.lorem.lines(rowCount)
        },
        data: () => ({
          edit: true
        })
      })

      expect(wrapper.find('textarea').attributes('rows')).toBe(String(rowCount))
    })
  })
})
