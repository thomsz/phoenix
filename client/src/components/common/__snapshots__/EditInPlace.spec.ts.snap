// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditInPlace should render form element: text 1`] = `
"<div class=\\"edit-in-place\\">
  <div class=\\"p-2 text-slate-500\\" data-testid=\\"interactive-element\\">Fake value</div>
</div>"
`;

exports[`EditInPlace should render form element: text 2`] = `
"<div class=\\"edit-in-place\\">
  <form name=\\"edit_in_place__fake_name\\"><input class=\\"text-slate-400 w-full h-full px-2 bg-transparent\\" name=\\"fake_name\\" type=\\"text\\"></form>
</div>"
`;

exports[`EditInPlace should render form element: textarea 1`] = `
"<div class=\\"edit-in-place\\">
  <div class=\\"p-2 text-slate-500 whitespace-pre-wrap\\" data-testid=\\"interactive-element\\">Fake value</div>
</div>"
`;

exports[`EditInPlace should render form element: textarea 2`] = `
"<div class=\\"edit-in-place\\">
  <form name=\\"edit_in_place__fake_name\\"><textarea class=\\"w-full p-2 bg-transparent\\" name=\\"fake_name\\" rows=\\"1\\">
      </textarea></form>
</div>"
`;
